File: week-3/requestment-2/py/functions/functions_test.py
from functions import *
import unittest

class TestFunctions(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestFunctions, self).__init__(*args, **kwargs)
        self.page = '''
          <a class="btn wide" href="/bbs/movie/index9501.html">‹ 上頁</a>
          <div class="title">
            <a href="/bbs/movie/M.1664272810.A.ED2.html">[負雷] Ｎ号棟鬧鬼 BLDG. N</a>
          </div>
          <div class="title">
            <a href="/bbs/movie/M.1664272810.A.ED2.html">[好雷]復仇之淵 槍戰很爽</a>
          </div>
          <div class="title">
            <a href="/bbs/movie/M.1664272810.A.ED2.html">[普雷] 行動代號：狼狩獵（首映）</a>
          </div>
        '''

    def testGetPage(self):
        url = 'https://www.ptt.cc/bbs/movie/index.html'
        page = getPage(url)
        self.assertNotEqual('', page)

    def testGetPreviousPageUrl(self):
        expected = 'https://www.ptt.cc/bbs/movie/index9501.html'
        actual = getPreviousPageUrl(self.page)
        self.assertEqual(expected, actual)

    def testGetTitles(self):
        expected = [
            '[負雷] Ｎ号棟鬧鬼 BLDG. N',
            '[好雷]復仇之淵 槍戰很爽',
            '[普雷] 行動代號：狼狩獵（首映）'
        ]
        actual  = getTitles(self.page)
        self.assertEqual(expected, actual)

    def testValidatePrefix(self):
        self.assertEqual(validatePrefix('[好雷]復仇之淵 槍戰很爽'), 0)
        self.assertEqual(validatePrefix('[普雷] 行動代號：狼狩獵（首映）'), 1)
        self.assertEqual(validatePrefix('[負雷] Ｎ号棟鬧鬼 BLDG. N'), 2)
        self.assertEqual(validatePrefix('[情報] 2022 第59屆金馬獎 入圍名單&入圍統計'), -1)

File: week-3/requestment-2/go/functions/functions_test.go
package functions

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func page() string {
	p := ""
	p += "<a class=\"btn wide\" href=\"/bbs/movie/index9501.html\">‹ 上頁</a>\n"
	p += "<div class=\"title\">\n"
	p += "  <a href=\"/bbs/movie/M.1664272810.A.ED2.html\">[負雷] Ｎ号棟鬧鬼 BLDG. N</a>\n"
	p += "</div>\n"
	p += "<div class=\"title\">\n"
	p += "  <a href=\"/bbs/movie/M.1664272810.A.ED2.html\">[好雷]復仇之淵 槍戰很爽</a>\n"
	p += "</div>\n"
	p += "<div class=\"title\">\n"
	p += "  <a href=\"/bbs/movie/M.1664272810.A.ED2.html\">[普雷] 行動代號：狼狩獵（首映）</a>\n"
	p += "</div>\n"
	return p
}

func TestGetResult(t *testing.T) {
	url := "https://www.ptt.cc/bbs/movie/index.html"
	page := GetPage(url)
	assert.NotEqual(t, "", page)
}

func TestGetPreviousPageUrl(t *testing.T) {
	expected := "https://www.ptt.cc/bbs/movie/index9501.html"
	actual := GetPreviousPageUrl(page())
	assert.Equal(t, expected, actual)
}

func TestGetTitles(t *testing.T) {
	expected := []string{
		"[負雷] Ｎ号棟鬧鬼 BLDG. N",
		"[好雷]復仇之淵 槍戰很爽",
		"[普雷] 行動代號：狼狩獵（首映）",
	}
	actual := GetTitles(page())
	assert.Equal(t, expected, actual)
}

func TestValidatePrefix(t *testing.T) {
	assert.Equal(t, ValidatePrefix("[好雷]復仇之淵 槍戰很爽"), 0)
	assert.Equal(t, ValidatePrefix("[普雷] 行動代號：狼狩獵（首映）"), 1)
	assert.Equal(t, ValidatePrefix("[負雷] Ｎ号棟鬧鬼 BLDG. N)"), 2)
	assert.Equal(t, ValidatePrefix("[情報] 2022 第59屆金馬獎 入圍名單&入圍統計"), -1)
}
